### PR TITLE
[SPARK-24127][SS] Continuous text socket source

### PR DESCRIPTION
**Pull request by arunmahadevan**
## What changes were proposed in this pull request?

Support for text socket stream in spark structured streaming "continuous" mode. This is roughly based on the idea of ContinuousMemoryStream where the executor queries the data from driver over an RPC endpoint.

This makes it possible to create Structured streaming continuous pipeline to ingest data via "nc" and run examples.

## How was this patch tested?

Unit test and ran spark examples in structured streaming continuous mode.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by arunmahadevan** [comment]
ping @jerryshao @tdas @jose-torres @HeartSaVioR for inputs.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
add to whitelist

**Comment by jose-torres** [comment]
I won't be able to look at this in detail until next week.

In general, I think this is a great source to have available. I wonder if it'd be worthwhile to try and abstract the record forwarding RPCs from here and ContinuousMemoryStream together.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+          val bufIndex = index - startOffset.offsets(partition)
+          val record = if (buckets(partition).size <= bufIndex) {
+            None
+          } else {
+            Some(buckets(partition)(bufIndex))
+          }
+          context.reply(
+            record.map(r => if (includeTimestamp) Row(r) else Row(r._1)
+            )
+          )
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala]:
```
context.reply(record.map(r => if (includeTimestamp) Row(r) else Row(r._1)))
```
Just in one line is ok I think.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+  }
+
+  override def commit(end: Offset): Unit = synchronized {
+    val endOffset = end.asInstanceOf[TextSocketOffset]
+    endOffset.offsets.zipWithIndex.foreach {
+      case (offset, partition) =>
+        buckets(partition).trimStart(offset - startOffset.offsets(partition))
+    }
+    startOffset = endOffset
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala]: If I understand right, this commit will never enter in your added test case.

**Comment by jerryshao** [comment]
I was thinking if it is too overkill to receive data in the driver side and publish them to the executors via RPC? This might give user a wrong impression that data should be received in the driver side and published to the executors again.

Just my two cents.

**Comment by jose-torres** [comment]
I think that's unavoidable if we want to have a socket source. The microbatch socket source has the same thing going on. I'd expect most people looking into implementation details of data sources will understand that they ought to read from executors in general.

**Comment by arunmahadevan** [comment]
A gentle ping for review @jose-torres , @jerryshao , @xuanyuanking

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ContinuousRecordEndpoint.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.spark.sql.sources.v2.reader.streaming.PartitionOffset
+
+case class ContinuousRecordPartitionOffset(partitionId: Int, offset: Int) extends PartitionOffset
+case class GetRecord(offset: ContinuousRecordPartitionOffset)
+
+/**
+ * A RPC end point for continuous readers to poll for
+ * records from the driver.
+ *
+ * @param buckets the data buckets
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ContinuousRecordEndpoint.scala]: nit: probably better to elaborate on what these are for, Seq[Seq[Any]] and Object aren't very informative types

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala [hunk 4]
```diff
@@ -256,7 +258,101 @@ class TextSocketStreamSuite extends StreamTest with SharedSQLContext with Before
...
+        "port" -> serverThread.port.toString).asJava))
+    reader.setStartOffset(Optional.empty())
+    val tasks = reader.planInputPartitions()
+    assert(tasks.size == 2)
+
+    val numRecords = 10
+    val data = scala.collection.mutable.ListBuffer[Int]()
+    val offsets = scala.collection.mutable.ListBuffer[Int]()
+    import org.scalatest.time.SpanSugar._
+    failAfter(5 seconds) {
```
Comment (jose-torres) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala]: Can't we use testStream for these tests?

Comment (arunmahadevan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala]: Probably we could use, but the `addSocketData` does not work for continuous source and thought the reader offsets could be validated better this way. (followed the approach in RateStreamProviderSuite)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ContinuousRecordEndpoint.scala (hunk 3 above)
Comment (arunmahadevan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ContinuousRecordEndpoint.scala]: Added more comments to clarify.

**Comment by arunmahadevan** [comment]
ping @tdas @jose-torres

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HeartSaVioR** [comment]
@arunmahadevan 
Sorry I forgot to review this so far. Could you fix merge conflicts? I'd pull the code to the local and review since the code diff is not small.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , rebased with master. 

ping @jose-torres @tdas @zsxwing for review.

**Comment by HeartSaVioR** [comment]
@arunmahadevan Thanks for rebasing. I'll take a look.

**Review by HeartSaVioR — COMMENTED** [review]
The code change looks good overall. 

Left comments mostly suggesting about deduplicating code between micro-batch and continuous and kind of defensive programming. Some comments could be shown as individual's preference and I'm definitely sure to follow Spark's preference.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , Addressed your comments. Let me know if I missed something. Also rebased and had to change more code to use the new interfaces. 

I hope if we can speed up the review cycles in general than leaving PRs to hibernation for a while and then the developer will loose the context and other things would have changed in the meanwhile.

**Comment by jose-torres** [comment]
The change looks broadly good (and important) to me. I'll defer to @HeartSaVioR wrt the in-depth review; let me know if there are any specific parts I should to take a look at.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+  initialize()
+
+  override def mergeOffsets(offsets: Array[PartitionOffset]): Offset = {
+    assert(offsets.length == numPartitions)
+    val offs = offsets
+      .map(_.asInstanceOf[ContinuousRecordPartitionOffset])
+      .sortBy(_.partitionId)
+      .map(_.offset)
+      .toList
+    TextSocketOffset(offs)
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala]: Yeah, agreed. I'm OK if same implications are used in other places.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+      } catch {
+        case e: IOException =>
+      }
+      socket = null
+    }
+    if (endpointRef != null) recordEndpoint.rpcEnv.stop(endpointRef)
+  }
+
+  private def initialize(): Unit = synchronized {
+    socket = new Socket(host, port)
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala]: Yeah you're planning to investigate and touch APIs then it sounds really good. May worth to file a new issue?

**Review by HeartSaVioR — APPROVED** [review]
LGTM given we are planning to tackle deduplication of codebase between micro-batch and continuous later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (arunmahadevan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousTextSocketSource.scala]: filed: https://issues.apache.org/jira/browse/SPARK-25000

**Comment by arunmahadevan** [comment]
retest this please

**Comment by arunmahadevan** [comment]
retest this please

**Comment by arunmahadevan** [comment]
retest this please

**Comment by arunmahadevan** [comment]
@HyukjinKwon this has been open for a while, would you mind taking this forward?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@jose-torres and @HeartSaVioR, is it good to go?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**